planner/planner.go

```diff
@@ -5,102 +5,122 @@ type Planner interface {
+	anyChanged := false
+
+	for _, rule := range p.rules[AnyKind] {
+		newNode, changed := rule.Rewrite(node)
+		anyChanged = anyChanged || changed
+		node = newNode
 	}
-	return nil, false
-}
 
-func (p LogicalToPhysicalPlanner) AddRules(rules []Rule) {
-	for _, rule := range rules {
-		p.rules[rule.Pattern().Root()] = rule
+	for _, rule := range p.rules[node.Kind()] {
+		newNode, changed := rule.Rewrite(node)
+		anyChanged = anyChanged || changed
+		node = newNode
 	}
+
+	return node, anyChanged
 }
 
-func (p LogicalToPhysicalPlanner) RemoveRules(rules []Rule) {
+func (p LogicalToPhysicalPlanner) AddRules(rules []Rule) {
 	for _, rule := range rules {
-		delete(p.rules, rule.Pattern().Root())
+		ruleSlice := p.rules[rule.Pattern().Root()]
+		p.rules[rule.Pattern().Root()] = append(ruleSlice, rule)
 	}
 }
 
 // Plan is a fixed-point query planning algorithm.
-// It enacts each rule on the query plan until no more rewrites are possible.
-func (p LogicalToPhysicalPlanner) Plan(root PlanNode) (PlanNode, error) {
-	var transformed bool
-	var newNode PlanNode
-
-	for _, rule := range p.rules {
-
-		// Try to match the root node
-		node, matched := p.matchRule(root, rule)
-
-		if matched && node == root {
-
-			if newNode, transformed = rule.Rewrite(node); !transformed {
-				// Record false match if root cannot be rewritten
-				p.falseMatch[node] = true
-			} else {
-				// Reassign root if successfully rewritten
-				root = newNode
-			}
-		}
-
-		for {
-
-			// No match means rule will never match
-			if node, matched = p.matchRule(root, rule); !matched {
-				break
-			}
-
-			if newNode, transformed = rule.Rewrite(node); !transformed {
-				p.falseMatch[node] = true
-			} else {
-				replacePlanNode(node, newNode)
+// It traverses the DAG depth-first, attempting to apply rewrite rules at each node.
+// Traversal is repeated until a pass over the DAG results in no changes with the given rule set.
+//
+// Plan may change its argument and/or return a new instance of QueryPlan, so the correct way to call Plan is:
+//     plan, err = planner.Plan(plan)
+func (p *LogicalToPhysicalPlanner) Plan(inputPlan *QueryPlan) (*QueryPlan, error) {
+
+	visited := make(map[PlanNode]struct{})
+
+	nodeStack := make([]PlanNode, len(inputPlan.Roots()))
+	copy(nodeStack, inputPlan.Roots())
+
+	for anyChanged := true; anyChanged == true; {
+		anyChanged = false
+		for ; len(nodeStack) > 0; {
+			node := nodeStack[len(nodeStack)-1]
+			nodeStack = nodeStack[0 : len(nodeStack)-1]
+
+			_, alreadyVisited := visited[node]
+
+			if ! alreadyVisited {
+				newNode, changed := p.matchRules(node)
+				anyChanged = anyChanged || changed
+				if node != newNode {
+					updateSuccessors(inputPlan, node, newNode)
+				}
+
+				// append to stack in reverse order so lower-indexed children
+				// are visited first.
+				for i := len(newNode.Predecessors()); i > 0; i-- {
+					nodeStack = append(nodeStack, newNode.Predecessors()[i - 1])
```

Comment:
What about successors? If a node (call it Node) is rewritten, there could conceivably be a rule that previously did not match a successor of Node but now does. Hence that successor should be visited again, right?